src/k8s/pkg/k8sd/features/contour/gateway.go

```diff
@@ -46,27 +46,48 @@ func waitForRequiredContourCommonCRDs(ctx context.Context, snap snap.Snap) error
 		return fmt.Errorf("failed to create Kubernetes client: %w", err)
 	}
 
-	return control.WaitUntilReady(ctx, func() (bool, error) {
-		resources, err := client.ListResourcesForGroupVersion("projectcontour.io/v1alpha1")
+	requiredCRDs := map[string][]string{
+		"projectcontour.io/v1alpha1": {
+			"contourconfigurations.projectcontour.io",
+			"contourdeployments.projectcontour.io",
+			"extensionservices.projectcontour.io",
+		},
+		"projectcontour.io/v1": {
+			"tlscertificatedelegations.projectcontour.io",
+			"httpproxies.projectcontour.io",
+		},
+	}
+
+	// checkRequiredCRDs checks if the required CRDs are present in the cluster.
+	checkRequiredCRDs := func(groupVersion string, required []string) (bool, error) {
+		resources, err := client.ListResourcesForGroupVersion(groupVersion)
 		if err != nil {
 			// This error is expected if the group version is not yet deployed.
 			return false, nil
 		}
 
-		requiredCRDs := map[string]bool{
-			"contourconfigurations.projectcontour.io":     true,
-			"contourdeployments.projectcontour.io":        true,
-			"extensionservices.projectcontour.io":         true,
-			"httpproxies.projectcontour.io":               true,
-			"tlscertificatedelegations.projectcontour.io": true,
```

Comment:
i see why we are doing this, but the changed code is now doing lots of memory allocations, and that's going to happen often, so let's try to avoid it.

You could keep the same code, by using `$group:$name` as your keys, e.g. `projectcontour.io/v1:tlscertificategelegations.projectcontour.io`.

Reply:
Great idea! Thanks :)